csharp/src/Microsoft.ML.OnnxRuntime/Tensors/Tensor.cs

```diff
@@ -313,9 +313,9 @@ protected Tensor(ReadOnlySpan<int> dimensions, bool reverseStride)
             long size = 1;
```

Comment:
(above) should you also allow dimensions of 0 length?

Reply:
I did discuss that with @jignparm. To me, we should support dimensions of length 0 because that is essentially a scalar. But @jignparm did say that there could be some sort "boxing up" of such scalars and also I did not find a model yet that broke on this. So, I would like to defer any changes in that line till we have a reason.

Reply:
The number of dimensions of the array (i.e. rank) can be 0 in Python.  These arrays are used to store scalar values, except that they are immutable, and cannot  be indexed into.

@hariharans29, does ORT allow 0-rank tensors in the base API? If not, then there the C# layer would never encounter such a tensor.

https://numpy.org/neps/nep-0027-zero-rank-arrarys.html

```python
X = np.array(99)
print(X)
print(X.shape)
print (type(X))
```

```
99
()
<class 'numpy.ndarray'>
```

Reply:
Sure @jignparm. Let me check that.

Reply:
@jignparm/@ericstj - AFAI can see, there is nothing in the base API layer that disallows creation of a scalar (0 dimensions), so I am guessing the C# layer must be modified to allow a scalar as well. Having said that, I would like to defer that change to another PR after 1.0 release in order to reduce the churn. Also, this PR is trying to solve - "Allow creation of empty tensors in C#" whereas allowing 0 dimension tensors is actually falling under a slightly different theme - "Allow creation of scalars in C#" (slightly different). FWIW - we would also have to modify another constructor (`protected Tensor(Array fromArray, bool reverseStride)` to allow creation of a "scalar" from a 1 element array. This will have to take an optional argument allowing creation of 0 length dimensions (as opposed to the current logic of creating dimensions of length 1). So this needs more discussion and triage.

Reply:
That seems reasonable.

Reply:
@hariharans29 -- thanks for checking into the support for 0-rank tensors in ORT. 

It sounds like 0-rank tensors are supported in the Python and C/C++ API, but not supported in the C# API currently. This is likely to cause a discrepancy in the future (with some rare model), but as you mentioned, it's a different change than the one required to unblock the Faster RCNN model.

Reply:
Correct :)

I think 0-rank tensors (scalars) and empty tensors are conceptually very different "special" tensors and hence let's deal with one problem alone in this change. Resolving this comment.